File: trunk/sabo/PointMass.py
from Vec2 import Vec2

class PointMass:
  def __init__( self, pos, mass = 1 ):
    self.pos = pos
    self.vel = Vec2()
    self.force = Vec2()
    self.mass = float( mass )
    self.friction = 0.1
    self.lastForce = Vec2()
    
  def applyForce( self, f ):
    self.force += f
    
  def update( self, dt ):
    acc = self.force / self.mass
    self.vel += ( acc - self.vel * self.friction ) * dt
    self.pos += self.vel * dt
    self.lastForce = self.force
    self.force = Vec2()

File: trunk/sabo/sabo/PointMass.cpp
#include <sabo/PointMass.h>

namespace sabo
{

  PointMass::PointMass( Body* body, const Vec2& pos, double mass ) :
    m_body( body ),
    m_pos( pos ),
    m_mass( mass ),
    m_friction( 0.25 ),
    m_maxSpeed( 10 )
  {
  }
  
  Vec2 PointMass::pos() const
  {
    return m_pos;
  }
  
  Vec2 PointMass::vel() const
  {
    return m_vel;
  }
  
  double PointMass::mass() const
  {
    return m_mass;
  }
  
  void PointMass::applyForce( const Vec2& force )
  {
    m_force += force;
  }
  
  void PointMass::update( double dt )
  {
    m_vel += ( m_force / m_mass - m_vel * m_friction ) * dt;
    if ( m_vel.length() > m_maxSpeed )
    {
      m_vel = m_vel.norm() * m_maxSpeed;
    }
    m_pos += m_vel * dt;
    m_force = Vec2();
  }

}
